### PR TITLE
Adding server dialog

### PR DESCRIPTION
**Pull request by catornot**
would be cool for servers to be able to create dialog boxes

### CODE REVIEW

**Comment by catornot** [comment]
I forgot to commit the changes for the mod.json

File: Northstar.Client/mod.json [hunk 1]
```diff
@@ -98,6 +98,17 @@
...
+		},
+		{
+			"Path": "ui/ui_server_dialog.nut",
+			"RunOn": "UI"
+		},
+		{
+			"Path": "cl_server_dialog.gnut",
+			"RunOn": "CLIENT",
+			"ClientCallback": {
+				"After": "serverDialogueOnClientInit"
```
Comment (F1F7Y) [inline Northstar.Client/mod.json:110]: Why is the S lowercase :(

**Review by GeckoEidechse — COMMENTED** [review]
I kinda feel like there's too much potential of abuse there compared to the server just using chat hooks for example.

**Comment by F1F7Y** [comment]
Just like I mentioned on the UMG PR, we could use some ruis ( weapon tips, the thing thylord uses in his sandbox, ... ) to allow servers to share info with clients without abusing `SendHUDMessage` ( Takyon >:( ) and spamming chat ( snnag >:( )
I agree with Gecko on the fact dialogs could be easily abused

File: Northstar.Client/mod.json (hunk 1 above)
Comment (catornot) [reply Northstar.Client/mod.json:110]: idk, I can fix it if you want

**Comment by catornot** [comment]
people just wouldn't play on servers that abuse dialogs, right?
I don't like chat hooks since they just cluster your screen.

**Comment by catornot** [comment]
might make weapon tips accessible server-side after this.

**Comment by GeckoEidechse** [comment]
> people just wouldn't play on servers that abuse dialogs, right?

Question, if server spams dialog boxes, can the player leave the server without having to resort to `ALT+F4`? Cause if not it's ripe for abuse.

**Comment by F1F7Y** [comment]
To chip in I was thinking something like this could work
image

```cpp
var welcome_message_rui = CreatePermanentCockpitRui( $"ui/fd_tutorial_tip.rpak", MINIMAP_Z_BASE )
RuiSetImage( welcome_message_rui, "backgroundImage", $"rui/menu/boosts/boost_holo_pilots" )
RuiSetString( welcome_message_rui, "titleText", "Welcome to cool server" )
RuiSetString( welcome_message_rui, "descriptionText", "Preferably the servers should be able to display ~3 messages so mods dont overwrite each other" )
RuiSetGameTime( welcome_message_rui, "updateTime", Time() )
RuiSetFloat( welcome_message_rui, "duration", 15.0 )
```
Credit goes to Thylord ( just copy pasted form his sandbox )

**Comment by F1F7Y** [comment]
Would be cool if we could also find a good looking rui for polls, so we dont use `SendHUDMessage`.

**Comment by ASpoonPlaysGames** [comment]
SendHUDMessage looks awful and it clashes with mods (you can only have one hud message at a time) so yeah an alternative would be nice

**Comment by F1F7Y** [comment]
I'm going to give this a shot myself :), if I get anywhere I'll make a PR to track it

**Comment by catornot** [comment]
Dialog is meant for polls and interactive stuff, it has 4 buttons.

**Comment by GeckoEidechse** [comment]
> Dialog is meant for polls and interactive stuff, it has 4 buttons.

I guess one we we could allow it, is if we limit it to certain states in the game, e.g. on round end to select next map as a vote or on player death. Additionally some form of "rate-limiting" check on the client to prevent the server from creating more than X dialogues in Y seconds.

My main concern is that server dialog is rather in your face IIRC and so receiving a message e.g. mid fight can get kinda annoying.

On a sidenote, @catornot can you maybe add a screenshot/screencap of it to your PR to show it in action? ^^

**Comment by catornot** [comment]
I will show screenshots when I get home ( Saturday )

**Comment by catornot** [comment]
image
screenshot :)

and a mod made with this.
[cat_or_not.Summoner.zip](https://github.com/R2Northstar/NorthstarMods/files/8934076/cat_or_not.Summoner.zip)
modified spawn mod.

**Comment by RoyalBlue1** [comment]
I just had an idea. To make a nicer version of this it is probably possible to use the shop ui from fd to something similar to this.
If you dont want do it ill probably do it after this is merged and I have some time

**Comment by catornot** [comment]
hmm, fd shop ui would be better for buttons since dialog can only have 4 buttons.

File: Northstar.Client/mod.json (hunk 1 above)
Comment (catornot) [reply Northstar.Client/mod.json]: fixed on this commit [22dd78d](https://github.com/R2Northstar/NorthstarMods/pull/380/commits/22dd78d4eb588058b479df5f06ac3a4c026013fb)

File: Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    args.append( dialog.image )
+    args.append( dialog.rightImage )
+    args.append( dialog.forceChoice.tointeger().tostring() )
+    args.append( dialog.noChoice.tointeger().tostring() )
+    args.append( dialog.noChoiceWithNavigateBack.tointeger().tostring() )
+    args.append( dialog.showSpinner.tointeger().tostring() )
+    args.append( dialog.showPCBackButton.tointeger().tostring() )
+    args.append( dialog.inputDisableTime.tostring() )
+    args.append( dialog.darkenBackground.tointeger().tostring() )
+    args.append( dialog.useFullMessageHeight.tointeger().tostring() )
```
Comment (ASpoonPlaysGames) [inline Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut:84]: Indentation problem?

File: Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+
+global struct ServerDialogData
+{
+	// var menu // cut no idea how this would work
+	string header = " "
+	string message = " "
+	// DialogMessageRuiData &ruiMessage
+	// array<int> messageColor = [161, 161, 161, 255] 
+	string image = "." // assets can not be sent to client without adding stuff to the launcher
+	string rightImage = "." // should be asset not string but ^
```
Comment (ASpoonPlaysGames) [inline Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut:33]: Does the new StringToAsset function help with this?

File: Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+
+	if ( dialog.buttonData.len() >= 3 )
+	{
+		args.append( StringReplace( dialog.buttonData[2].text, " ", "./s.", true, true ) )
+		_addPlayerButtonCallBack( player, dialog.buttonData[2].callback )
+	}
+
+    if ( dialog.buttonData.len() >= 4 )
+	{
+		args.append( StringReplace( dialog.buttonData[3].text, " ", "./s.", true, true ) )
```
Comment (ASpoonPlaysGames) [inline Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut:108]: Indentation problem?

File: Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+}
+
+void function SendMessageServerDialog( entity player, string message )
+{
+	ServerDialogData dialog
+	dialog.message = message
+	SendServerDialog( player, dialog )
+}
+
+void function AddDialogButton( ServerDialogData dialog, string text , void functionref(entity) callback )
```
Comment (ASpoonPlaysGames) [inline Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut:132]: unnecessary space after `string text`

Comment (ASpoonPlaysGames) [reply Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut:132]: Also idk if we like to put spaces in `functionref( entity )` or if we prefer `functionref(entity)`

File: Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    float inputDisableTime = args[9].tofloat()
+    bool darkenBackground = args[10].tointeger() == 1
+    bool useFullMessageHeight = args[11].tointeger() == 1
+
+    string button0Text = StringReplace( args[12], "./s.", " ", true, true )
+    string button1Text = StringReplace( args[13], "./s.", " ", true, true )
+    string button2Text = StringReplace( args[14], "./s.", " ", true, true )
+    string button3Text = StringReplace( args[15], "./s.", " ", true, true )
+
+    RunUIScript( "ClientCallBack_RequestDialog", header, message, image, rightImage, forceChoice, noChoice, noChoiceWithNavigateBack, showSpinner, showPCBackButton, inputDisableTime, darkenBackground, useFullMessageHeight, button0Text, button1Text, button2Text, button3Text )
```
Comment (ASpoonPlaysGames) [inline Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut:201]: Might be better to split this into multiple lines, for readability, although I'm not sure if squirrel likes this

File: Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+{    
+    DialogData dialogData
+
+    dialogData.header = header
+    dialogData.message = message
+
+    // if ( image.find("/") != null )
+    //     dialogData.image = asset( image ) // this doesn't work and using compilestring( "return $\"" + image + "\"" )() is not safe
+    // if ( rightImage.find("/") != null )
+    //     dialogData.rightImage = asset( rightImage )
```
Comment (ASpoonPlaysGames) [inline Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut:216]: new StringToAsset function could help with this potentially?

File: Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut [hunk 8]
```diff
@@ -0,0 +1,259 @@
+#if SERVER
+global function ServerDialogInit
+global function SendServerDialog
+global function SendMessageServerDialog
+global function AddDialogButton
+#elseif CLIENT
+global function ServerDialogueOnClientInit
+#elseif UI
+global function ClientCallBack_RequestDialog
```
Comment (ASpoonPlaysGames) [inline Northstar.Custom/mod/scripts/vscripts/sh_server_dialog.gnut:9]: This should probably be called `ClientCallback_RequestDialog` (or dialogue if we are feeling particularly british ig) as Callback is one word

**Review by ASpoonPlaysGames — CHANGES_REQUESTED** [review]
Couple of formatting things mostly, the logic seems sound though

**Comment by GeckoEidechse** [comment]
@catornot or @ASpoonPlaysGames if any of the code review feedback by Spoon above has been addressed, make sure to click on  "Resolve Conversation" on that part, otherwise GitHub blocks any option of merging ^^

**Comment by GeckoEidechse** [comment]
Also I would still appreciate some more feedback by other modders in regards to whether this is needed (e.g. for some specific mod to have better UX) or whether the potential for abuse is too big here (spamming server dialog messages to player so they cannot play).

**Comment by Alystrasz** [comment]
This could be used with voting mods to provide players with "you voted for the amogus map" feedback = good UX.
Also, if a mod abuses this system, the TS council should warn its author, and remove said mod from store if no action is taken :)

**Comment by uniboi** [comment]
I don't see how this could be more disruptive to gameplay than pinning a player to some prop or something. If players feel like a server is using them too often, they're going to play on another server

**Comment by EladNLG** [comment]
Small suggestion - use the SP choice prompts.

**Comment by GeckoEidechse** [comment]
So according to @EladNLG this blocks all input (apart from `F1`-`F12` keys) making it impossible to leave a server via `Esc` -> `Quit game` when the server is abusing this feature.

Players should always be able to leave a server in a standard manner without having to resort to using console commands or force closing the game.
Therefore I'm gonna close this PR.

If you can make it work so that players can still leave a server that is abusing this (or if I'm simply mistaken and it's possible to leave server with server dialog getting spammed at client) feel free to re-open. ^^

A better way to do polls for e.g. map votes than the current chathook implementation is still wanted. So some system that allows for e.g. easier map votes while still fulfilling the requirements mentioned above is something we wanna add to Northstar <3

**Comment by catornot** [comment]
I added one cooldown to dialogs and a check for active menus

https://user-images.githubusercontent.com/41955154/184451361-e1bd0916-e10d-41db-a3ef-c954c3f3616a.mp4

**Comment by GeckoEidechse** [comment]
> problem with this RUI is that it only exposes two answer choices, and no question text

The question text could come from another RUI element to remedy that issue. The only 2 choices thing is annoying though :/

**Pull request closed without merging**